### PR TITLE
Runescape-y chat on the game world

### PR DESCRIPTION
**Pull request by ShiftyRail**
Ported from Eris (with permission from TG).

# About the PR

This PR adds a system to display things heard by mobs on the game map, very much like in the classic game Runescape.

# FAQ

- Is it a toggle ?
Yes. It's currently off by default. If not enough people wish it enabled, it can be removed from players without too much hassle. You will have three different setting to edit: have it enabled at all, the maximum number of characters you want it to display, and if you wish to have it enabled for only mobs, or for everyone in the game world. 
More advanced options (such as making the font smaller, or trying a different typeface) can be done editing your `skin.dmf`.

- How is the colour of the text generated ?
The font colour is different depending on the **name** of mob speaking, plus a random number variying round after round. This means you can still impersonate someone, but you won't keep a persistent colour across rounds.

- How is this PR different from the TG one ?
I made the text slightly smaller than the setting used on TG. Unlike TG, the text does not get boldface when you are yelling using a megaphone. Whispering does not make the maptext display italics, it just make it a bit smaller.

- How bad in terms of performance ?
It is very light for the server, as it's nothing more than a few more checks on hearing. The icon displaying and animation is done entirely from the client, and it's not really that taxing.

- How abusable it is ?
The overlay moves with you and anything you say is laid over what you are already saying. By default, the maximum "space" taken by the overlay is 3x2 tiles. The overlay does not catch any clicks, and disappears in about 4 seconds.

biggest space

# Gallery

maybe or maybe not
recorders
shuttle

If we ever decide to use 64x64 base sprites, this text can be made a lot smaller.

maptext64x

### CODE REVIEW

**Comment by DamianX** [comment]
#26632 also defines migration number 23, so be careful.

**Comment by Dacendeth** [comment]
does this work with Me or just Say?

**Comment by ShiftyRail** [comment]
@Dacendeth Just on say, currently.

**Comment by Kammerjunk** [comment]
add shortcuts to certain key ("most-used") phrases like
- "Are you valid?"
- "I am not valid."
- "He is valid."
- "Where are the valids?"

**Comment by ShiftyRail** [comment]
This is an excellent idea, Kammer, I'll do that in a later PR.

**Comment by Kammerjunk** [comment]
thank you
I think there are a lot of good ideas that can stem from this PR:
image

**Comment by ShiftyRail** [comment]
Why don't we add a toggle which makes people post good faith suggestions?

**Comment by SonixApache** [comment]
this can take priority on migration 23, no worries

**Comment by Kammerjunk** [comment]
"good faith," at times, is a meme word that people use because they don't understand that it isn't the same as mockery. that picture is obviously the latter.

"just make it a toggle" isn't a solution in a social roleplaying game because the effect isn't individual. I don't want to play a MOBA and I don't want to play with people who are playing a MOBA. 
would you agree to add a toggle to switch the chat to how it currently looks in Unity or SS14? the toggle is for one player but it means you have people playing SS13 and people playing Runescape13 and they're connected to the same server. that's what that comment in the thread (which I think you think I made) is saying.
in short, it's the same argument as "if you don't like it, just don't use it"

this looks like a meme PR.

**Comment by ShiftyRail** [comment]
This isn't a meme PR, this is a real content PR I worked hard on, and Pomf suggested that I did it, even telling me he got the authorisation from TG to port it.
I think it looks cool.

Am I to understand you don't want the TG toggle, after all ?

**Comment by Kammerjunk** [comment]
I'm not really sure what you might understand

**Comment by ShiftyRail** [comment]
That you want a TG toggle, of course, why would you post that you want a TG toggle if you didn't want a TG toggle?

**Comment by Kammerjunk** [comment]
if you aren't willing to or capable of defending your PRs, why are you a collab?

**Comment by ShiftyRail** [comment]
I'm willing and capable to defend this PR. It's content, it's a toggle, and it's something people seem to like, and that Pomf encouraged me to code.

image

As for the TG redirect PR, I have it on record that you want it, so consider it done on request.

**Comment by Kammerjunk** [comment]
you aren't
it's bad content and bad content gets the "it's a toggle, don't use it if you don't like it" defence which I addressed above

forcing a bad joke that hard when you don't get the response you're looking for just makes it worse

**Comment by ShiftyRail** [comment]
Then say you don't like it, make a downvote, and say, "It's a TG feature and I don't like it"
Don't make an epic witty unreadable irony shitpost

You're perfectly allowed to think it's bad content, I think it's worth trying as a toggle and it's revertable if nobody likes it

**Comment by Kammerjunk** [comment]
I did all of that. don't blame your crippling lack of experience with human communication on me

**Comment by Kharkov72** [comment]
I just wish there was a better way than having it cover up tiles

**Comment by Kammerjunk** [comment]
there is. it takes up most of the right half of your screen.

**Comment by ShiftyRail** [comment]
There's a better way to do this, it's 64x64.

**Comment by kilozombie** [comment]
This is also done on Eris. Just say you're porting it from there and not TG and nerds can't complain

**Comment by ShiftyRail** [comment]
By the way, Travis fails because we need to bump it to 513.

**Comment by ShiftyRail** [comment]
maxlength
news
shuttle

Reduced the max message length (from 110 chars to 68; you can still say "ONE DAY WHILE ANDY WAS ..." fully, don't worry), removed the black outline, made objects in the world display a smaller font so that it's less invasive, and made the messages semi-transparent (on the level of a window) if you want to see below them.

The messages will get smaller the further away from you the speaker is.

You can edit the size and the outline in your `skin.dmf` if you want.

**Comment by RaccoonBro** [comment]
Removing the black outline by default utterly fucks readability, like what does this even say
image

**Comment by ShiftyRail** [comment]
I could make it a verb in game if you don't want to fuck with your .dmf

**Comment by RaccoonBro** [comment]
Would probably be for the best

**Comment by Exxion** [comment]
If you want to be fancy you could probably use `BLEND_MODE` and a color matrix to make the text be a very different color from whatever is below it

EDIT: Or you could just give the text a background, which would be easier and better-looking

**Comment by tuypo1** [comment]
Sounds like a great idea, when in a busy situation where you have to manage lots of stuff being able to see exactly who is saying something and from where at a glance is great

**Comment by ghost** [comment]
I think the heart of Kammerjunk's comments that it will change how the game is played is true. If this does get coded then by all means make it a test week and see how it functions **without** a toggle out of it, so that the sample size of players who get to see what it feels like is as large as possible, then make a poll about whether it should be included at all.

UI is a large part of the experience of this server and we should be careful when making such a large change.

Now for an actual question; how do radios work? I don't play on other servers because I am faithful to my /vg/ overlords.

**Comment by ShiftyRail** [comment]
Why have it work without a toggle ? There's absolutely no reason for it to be a "all or nothing" feature, it's the kind of thing that would divide the community and only serves the interests of the most extremist factions.
It's exactly the same thing as attack animations, or even the typing indicator (now without server input lag), some QoL which does change the game a bit but not fundamentally speaking.

Making it something you **have** to have enabled is a sure way for it to fail and cause frustration further down the line. If by any chance this has to go through as a "all or nothing" feature, then have attack animations, typing indicator, progress bars, whatever you think of also be polled and t if the poll fails.

>How do radios work ?

I've changed it from TG so that the speech you hear from radios stays in goonchat only. This can be changed if the initial reception to runechat isn't bad.

**Comment by ghost** [comment]
Shifty, I mean for the test week. Read what I said:

> so that the sample size of players who get to see what it feels like is as large as possible

And its less so because the game relies upon conversation more than valids at the end of the day. It's going to be a contentious feature and I think everyone should experience what its like if it gets introduced rather than just the few who think to turn on their preferences to it or even know it exists at all.

Edit: and thank you, does that mean with radios that if it were integrated they'd appear only for you next to you/the radio's location?

**Comment by ShiftyRail** [comment]
I don't like the idea of forcing it on players in any way. It's a sure way to get a knee-jerk reaction and a downvote because something was done without them being informed.

You could in the future implement that messages sent through headsets have a special font or icon to them (this is what TG did). It is technically possible to display messages you hear through your headset with this PR, but I believe this wouldn't work at all.

**Comment by ghost** [comment]
Unfortunately without a hard toggle and a proper test I'd likely be on kammerjunk's side here. I'm a fan of MUDs for the way they present text helping them be a text based game, which many people would call outdated because they don't work like more modern RPGs. Maybe I'm just an old man and the server will move on without me, in which case I'll be a sad old man clinging to what spess used to be like.

Also the more visible text is better, clearly.

**Comment by ShiftyRail** [comment]
Let's assume we have 10% old men clinging to their traditions, 10% young and vivacious people wanting the change, and 80% not caring one way or the other. A forced test will only amplify the voice of extremists on either side.

I'm very doubtful of a "let's force it and see how it goes at the end poll" since that poll with antag objectives which was intentionally designed to make one side look like it had more support. A serene test is an opt-in one with a clear question at the end.

**Comment by strawberrytan** [comment]
Runescape chat and 64x64 when

It actually reminds me of how project zomboid handles proximity chat

**Comment by Kammerjunk** [comment]
I'm really not seeing your logic there. however many people don't care one way or the other doesn't matter because they presumably won't vote one way or the other, so it's the remaining percentage of people whose (hopefully informed?) opinions you care about. if they can't form one because they have no experience, how will you know what they think?
someone making a rigged poll isn't a good argument against anything that isn't making a rigged poll. make a poll that isn't bad.

also see above why "just toggle it" is bad and dumb

**Comment by ShiftyRail** [comment]
Well yeah, it'll impact your game since people have it on might play the game differently. The goal of a PR is to impact your game. I don't see any argument that it would impact your game *negatively*. Not in the first comment you made at any rate.

It'll impact your or anyone's game a lot less if you personally don't have it enabled on you by force. Which is the condition needed for a serene testing. A serene testing with people choosing to try the feature and seeing how it affects the game rather than having it forced on them one way or the other by people who give in to pre-made opinions.

And I still maintain that it's **a good thing for the game**. _It makes the world feel more alive_, make you react more to things happening around you, _and get a quick view of what people are doing at a glance_. It's a great feature which _helps with roleplay_, _as it makes it easier to see what multiple people are saying instead of giving up,_ _immersion_, as it makes the _small talking objects in the world a lot more noticeable,_ and overall moves away from the tedium, boring, and less fun parts of the game to something which actually gets people involved in roleplay (especially in busy situations, like a trial, or something) in a user-friendly way.

It's really great for the game. I don't understand at all the inherent hostility this can cause. I'm not really willing to try and give more thought and understand the mindset of someone who comes with a "mockery". You don't mock people if you want to convince them they're on the wrong path ; and you cannot tell them they do not understand language if you don't try to communicate clearly.

In short, don't make a trash comment if you don't want your comment and all subsequent ones to be treated like trash. You're kinda forcing people into taking a defensive and non-communicative stance if you come up with le epic passive-aggresive irony.

I don't really see clearly how people having a better picture  of what is happening around them makes it a bad thing for (you)r spacemen experience, Even if it did, it looks like there's at least enough support for a try, maybe even a majority wanting the change. Having it as a toggle let you reduce the effects it would have on your game if it were to be merged.

**Comment by Cisrox** [comment]
this mechanically shifts player focus from the text wall to the game window, which is major and personally i am open to atleast trying it, i think it has potential to make the game more fun

**Comment by ComradeDispenser** [comment]
I think it will be a positive change, when you can actively see people talking even when you're focused on the game window instead of keeping your eyes constantly glued to the side the of the screen where all the text shows up you are allowed to be more involved in actually playing the game while also making it easier to see who's talking and who's saying what, I think it will make conversations and player interactions easier to follow and probably even quicker because the text will be easier to read on the fly.

I feel like this would only be a good thing for people who are interested in actually roleplaying or at least doing anything other than silent shuffling because it just makes conversation in general easier to follow and take in.

**Comment by Dacendeth** [comment]
Do we really need the chat bubble overhead anymore if you have this option on? Not sure how hard that would be to have turn off while you had this option on.

**Comment by ComradeDispenser** [comment]
Chat bubble let's you see someone is typing so it would still be useful.
I dont see any reason we shouldn't have both.

**Comment by ShiftyRail** [comment]
I'll add some toggle creep, do some runtime hunting, and then this should be good2go.
If this is merged, I heavily encourage people to voice how they feel during the trial period.

**Comment by Kammerjunk** [comment]
everything you've said re. testing it is a complete non sequitur and I'm not sure if you've actually considered it. if you want opinions that aren't pre-made, you need people to be exposed to it. you need it to "impact their game."
any other set of circumstances results in less developed opinions by average because people have no or little experience with it.

as far as game feel goes, the world feels plenty alive as it is and there is no lack of indicators for activity or action.
as far as roleplaying is concerned, I don't understand e.g. the view that having a perfect picture (were this to give that) of an area with many ongoing conversations or many people speaking at the same time is an improvement to roleplay. it's neither an increase to immersion nor to realism. focus doesn't work that way. busy situations are chaotic.

everything after that is defensive padding to your post that doesn't address anything. if you don't understand the hostility I showed you in the coderclub when you consistently ignored exactly what I was honestly saying to you in favour of telling me that you don't like people being dishonest, I don't know how to help you. considering the way you behaved, you can't excuse it as being intentional either.

see above for why turning everything to a toggle, especially things to do with communication, isn't a very good solution in a game with heavy focus on both mutual and incomplete communication.

**Comment by SonixApache** [comment]
i like this PR and the way it was implemented in the OP

**Comment by ShiftyRail** [comment]
image

Some sort of toggle creep which was requested. There's now a clear distinction between "objects in the world", and "mobs in the world". Objects in the world can be made entirely mute from goonchat if you want, through I have to check if this doesn't accidentally kill objects speaking through a radio. It shouldn't, as the speaker is a virtual hearer, but hey.

If you are weary of spam, here is what a macro+REP spam looks like in goonchat and runechat, given that I upped TG's limitations on screensize occupied by the thing.

https://streamable.com/v152mm - the macro on Goonchat
https://streamable.com/2hyzbs - the same macro on runechat.

In both cases the 'anti-spam' kinda fails, for Goonchat there's so many messages sent that the script to check if the last messages are identical bugs out, for Runechat, BYOND Client simply cannot process so many lines.

**Comment by bathosbathtime** [comment]
I like the idea of this being used for machines, consoles, and other environmental objects.
I dislike the idea of this being used for player speech as it will influence how they communicate in-game. This is a potential issue that can not be solved simply with a toggle as if it is on by default as part of the problem is giving new players a more "game"-y impression of the game from its chat. Even if it is not, the users that do enable it will change how they speak to better fit the Runescape chat.
I do agree that this will help the game feel more alive, and that it is overall a good system. I just fear that this will change the behavior of the players that use it and the feeling of the game in a way that a toggle does nothing to solve.
While this pull request does have overwhelming support to the point where you wouldn't need to compromise to get it through, I would much rather have it apply to objects only. If you do not find that agreeable, that is fine. I will simply have to get used to the mob chat existing in spess. I hope to find Kammer's and my fear to be unfounded, if that is the case.

File: __DEFINES/planes+layers.dm [hunk 1]
```diff
@@ -195,6 +195,7 @@ What is the naming convention for planes or layers?
 	#define ABOVE_SINGULO_LAYER 	7
 	#define GRAVITYGRID_LAYER 		8
 	#define SNOW_OVERLAY_LAYER		9
+	#define CHAT_LAYER 12.1
```
Comment (DamianX) [inline __DEFINES/planes+layers.dm:198]: why is this value so weird?

Comment (Exxion) [reply __DEFINES/planes+layers.dm:198]: Probably what it was on TG
I'm guessing they have a lot more layers than we do
Makes sense considering how many things we have sharing layers

Comment (ShiftyRail) [reply __DEFINES/planes+layers.dm:198]: Yeah, they needed it to be above mobs but below lights.

File: code/datums/chat_message.dm [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+	var/h_int = round(h/60) // mapping each section of H to 60 degree sections
+	var/c = (1 - abs(2 * l - 1)) * s
+	var/x = c * (1 - abs((h / 60) % 2 - 1))
+	var/m = l - c * 0.5
+	x = (x + m) * 255
+	c = (c + m) * 255
+	m *= 255
+	switch(h_int)
+		if(0)
+			return "#[num2hex(c, 2)][num2hex(x, 2)][num2hex(m, 2)]C8"
```
Comment (DamianX) [inline code/datums/chat_message.dm]: you should be able to replace these with 
```suggestion
			return "#[rgb(c, x, m)]C8"
```
I remember testing this a while back and it was a lot faster

**Comment by ShiftyRail** [comment]
Tested & good to go

**Comment by RaccoonBro** [comment]
This is hidden when people talk over the radio, right?

File: code/modules/client/client procs.dm [hunk 3]
```diff
@@ -546,3 +546,9 @@ NOTE:  You will only be polled about this role once per round. To change your ch
 /client/proc/handle_hear_voice(var/mob/origin)
 	if(prefs.hear_voicesound)
 		mob.playsound_local(get_turf(origin), get_sfx("voice"),50,1)
+
+/client/verb/toggle_runechat_outlight()
```
Comment (Kurfursten) [inline code/modules/client/client procs.dm]: outlight?

**Comment by tuypo1** [comment]
Do you see anything on the game window if you can tell somebody is talking but not what they are saying such as from being deaf or too far away

File: code/modules/client/client procs.dm (hunk 3 above)
Comment (ShiftyRail) [reply code/modules/client/client procs.dm]: Ooops

**Comment by ShiftyRail** [comment]
@RaccoonBro You do see the chat text if they are talking over an (unsecured) radio and you are in view. You will not see a chat bubble if you are not in view and hear something over the radio, and you'll not see something if they're talking into the security headset (for instance), which is considered secured.
I'll make it so a small icon appears if this is the case so that you wouldn't confused ambient talk for something else.

@tuypo1 Nothing appears if you're not hearing anything.

**Comment by ShiftyRail** [comment]
image

Here's how it looks like when someone talks into a radio.

**Comment by Exxion** [comment]
You can normally only hear someone talking into a headset if you're on an adjacent tile, meaning there's an incentive to use this over the chat window. This is not good.

**Comment by ShiftyRail** [comment]
If you can't see the message on the goonchat window, there will be no text on the map. 

The information used to decide whether or not the mob hears the message is the same, the `Hear()` proc, with some additional checks normally handled by ` show_message()`, mostly the fact the image will not be created if the mob is deaf or unconscious, so the text message will not be shown if you're not adjacent. 
The text message to be displayed to Goonchat and the image to be displayed by the client are sent at the same time.

It *does* make it more obvious that they talked into their headset with the icon, but this information is already displayed on the goonchat window because the text appears twice and in italics.

In the same way, if someone talks, and then moves in view of another player while he can see his message floating over his head, the other mob will NOT see it if he didn't hear the message (and had it displayed in Goonchat) in the first place. Images are client-side.

**Comment by Exxion** [comment]
You said in view before. If it's actually within the same range, then it's fine.
